### PR TITLE
statistics: fix the bug that cold peer stat won't be updated

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Currently, for hot read peer stat reporting, it is depended on the store heartbeat. If the peer become cold after it is marked as hotPeerStat, the store heartbeat won't take it which make the PD can't automatically update cold peer stat.

### What is changed and how it works?

Check cold peer stat during each store heartbeat to fix this problem.

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: server/statistics/hot_cache.go [hunk 1]
```diff
@@ -17,11 +17,17 @@ import (
 	"context"
 
 	"github.com/tikv/pd/server/core"
+	"go.uber.org/atomic"
 )
 
 // Denoising is an option to calculate flow base on the real heartbeats. Should
 // only turned off by the simulator and the test.
-var Denoising = true
+var Denoising = atomic.NewBool(true)
```
Comment (rleungx) [inline server/statistics/hot_cache.go:25]: I think it's unnecessary to use an atomic variable?

Comment (lhy1024) [reply server/statistics/hot_cache.go:25]: there may be race if no atomic.

File: server/statistics/hot_cache.go [hunk 2]
```diff
@@ -17,11 +17,17 @@ import (
 	"context"
 
 	"github.com/tikv/pd/server/core"
+	"go.uber.org/atomic"
```
Comment (JmPotato) [inline server/statistics/hot_cache.go:20]: Why not just using `sync/atomic` pkg here?

Comment (Yisaer) [reply server/statistics/hot_cache.go]: `sync/atomic` also lgtm.

File: server/statistics/hot_cache.go (hunk 1 above)
Comment (Yisaer) [reply server/statistics/hot_cache.go]: we met data race before

**Review by lhy1024 — APPROVED** [review]

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -554,10 +556,11 @@ func (c *RaftCluster) HandleStoreHeartbeat(stats *pdpb.StoreStats) error {
 		}
 		peerInfo := core.NewPeerInfo(peer, 0, 0,
 			peerStat.GetReadBytes(), peerStat.GetReadKeys(), interval)
-		item := statistics.NewFlowItem(peerInfo, region, nil)
+		item := statistics.NewPeerInfoItem(peerInfo, region)
 		c.hotStat.CheckReadAsync(item)
 	}
-
+	coldItem := statistics.NewColdItem(storeID, regionIDs, interval)
```
Comment (lhy1024) [inline server/cluster/cluster.go:562]: `coldItem` is confusing, we need a more fit name.

Comment (Yisaer) [reply server/cluster/cluster.go]: updated.

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -554,10 +556,11 @@ func (c *RaftCluster) HandleStoreHeartbeat(stats *pdpb.StoreStats) error {
 		}
 		peerInfo := core.NewPeerInfo(peer, 0, 0,
 			peerStat.GetReadBytes(), peerStat.GetReadKeys(), interval)
-		item := statistics.NewFlowItem(peerInfo, region, nil)
+		item := statistics.NewPeerInfoItem(peerInfo, region)
 		c.hotStat.CheckReadAsync(item)
 	}
-
+	coldItem := statistics.NewUnReportStatsHandle(storeID, regionIDs, interval)
```
Comment (lhy1024) [inline server/cluster/cluster.go:562]: `coldItem` also need to be updated

File: server/statistics/hot_cache.go [hunk 5]
```diff
@@ -33,19 +48,51 @@ type HotCache struct {
...
+// NewPeerInfoItem creates FlowItem for PeerInfo
+func NewPeerInfoItem(peerInfo *core.PeerInfo, regionInfo *core.RegionInfo) *FlowItem {
 	return &FlowItem{
-		peerInfo:    peerInfo,
-		regionInfo:  regionInfo,
-		expiredStat: expiredStat,
+		peerInfo:            peerInfo,
+		regionInfo:          regionInfo,
+		expiredStat:         nil,
+		unReportStatsHandle: nil,
```
Comment (lhy1024) [inline server/statistics/hot_cache.go:85]: how about `unreportStatCollect`？

**Review by rleungx — APPROVED** [review]

File: server/statistics/hot_peer_cache.go [hunk 6]
```diff
@@ -433,33 +409,24 @@ func (f *hotPeerCache) updateHotPeerStat(newItem, oldItem *HotPeerStat, deltaLoa
 	isFull := newItem.rollingLoads[0].isFull() // The intervals of dims are the same, so it is only necessary to determine whether any of them
 	if !isFull {
 		// not update hot degree and anti count
-		newItem.HotDegree = oldItem.HotDegree
-		newItem.AntiCount = oldItem.AntiCount
+		succeedItemDegree(newItem, oldItem)
```
Comment (lhy1024) [inline server/statistics/hot_peer_cache.go:412]: how about `inheritItem`

**Pull request merged**